packages/grafana-data/src/types/datasource.ts

```diff
@@ -374,6 +375,12 @@ abstract class DataSourceApi<
    * @alpha -- experimental
    */
   getDefaultQuery?(app: CoreApp): Partial<TQuery>;
+
+  /**
+   * Optionally, use this method to migrate a query
+   * @alpha -- experimental
+   */
+  migrateQuery?(query: TQuery): Promise<TQuery> | TQuery;
```

Comment:
one thing to know @sunker is that this could be considered a breaking change. If a plugin implements a `migrateQuery` function but it has a different signature, it will throw a syntax error when building. Also, if the plugin doesn't expect this function to be automatically executed it could cause trouble. Thoughts? Have we worried about this before?

Reply:
yep, levitate complains about this :)

Reply:
Hmm good question. Still we _must_ be able to add new methods. Thoughts @mckn? 

If you remove the `migrateQuery` signature from DataSourceApi and only define it in DSWithBackend it should at least limit the impact a bit. Then in the HOC you'll need the following check:`dataSource instanceof DataSourceWithBackend`.

Reply:
That would silence levitate but yes, the breaking change will still be present for plugins using DSWithBackend. In any case, this is probably useful for plugins without a backend, no? We may want to leave this as it is so frontend-only plugins can also define their `migrateQuery`.

Reply:
I have also removed the `migrateQuery` implementation from the `DatasourceWithBackend` and moved it to a utils function so for plugins to opt-in on it, they will need to do:

```
  migrateQuery(query: MyQuery): Promise<MyQuery> | MyQuery {
    return postMigrateQuery(query);
  }
```

This is, so the check `if (props.query && props.datasource.migrateQuery) {` below makes sense (if not it's always defined for `DatasourceWithBackend`). That way we also avoid experimentanl code in `DatasourceWithBackend`.

Reply:
Another reason to introduce DSWithAPIServerBackend. :) 

True that this would be useful for frontend-only data sources, but then we'd need to also update the query runner so that the migration is executed also before calling the query method. 

Speaking of that, how does that work with the admission handler? I assume the backend migration is somehow called as part of the query execution in the backend, but what about the frontend? Is the developer expected to handle legacy query structure in for example datasource.filterQuery and datasource.query?

Reply:
in the query execution path, the migration step is not automatic. Plugins need to call it explicity. This is because of performance reasons, since running the migration before `QueryData` and then call that function would imply an additional marshalling and unmarshalling. See how it works in the example: https://github.com/grafana/grafana-plugin-examples/blob/01b8d2fe30c4c0a8e2b42e8fc37489d35755e168/examples/datasource-http-backend/pkg/plugin/datasource.go#L179

If we follow the same approach, we could document that for doing migrations in frontend-only datasources, people will need to also manually call that.